File: practice/11_Container_With_Most_Water.kt
fun main() {
	println(container(intArrayOf(1, 8, 6, 2, 5, 4, 8, 3, 7)))
    println(container(intArrayOf(1, 1)))
    println(container(intArrayOf(1, 2, 4, 3)))
}

fun container(height: IntArray): Int {
	var max = 0; var i = 0; var j = height.size-1;
	
	while(i<j) {
		val min = minOf(height[i], height[j])
		val width = j-i
		max = maxOf(max, min*width)

		if(height[i] <= height[j]) {
			i++
		} else {
			j--
		}
	}

	return max;
}

File: practice/11_Container_With_Most_Water.js
function mostWater(height) {
    let max = 0,
        i = 0,
        j = height.length - 1;

    while (i < j) {
        const min = Math.min(height[i], height[j]);
        const width = j - i;
        max = Math.max(max, min * width);
        if (height[i] <= height[j]) {
            // let temp = height[i] * (j - i);
            // max = Math.max(max, temp);
            i++;
        } else {
            // let temp = height[j] * (j - i);
            // max = Math.max(max, temp);
            j--;
        }
    }

    return max;
}

// console.log(mostWater([1, 8, 6, 2, 5, 4, 8, 3, 7]));
// console.log(mostWater([1, 1]));
console.log(mostWater([1, 2, 4, 3]));
